syntax = "proto3";

enum ClusterRole {
    UNSET = 0;
    LEADER = 1;
    FOLLOWER = 2;
}

enum ClusterState {
    INITIALIZE = 0;
    POPULATING = 1;
    IDLE = 2;
    EXECUTING = 4;
    ERROR = 8;
}

enum ClusterMessageType {
    UNKNOWN = 0;
    ACK = 1;
    SIGNAL_HOT_RELOAD = 2;
    ANNOUNCE = 4;
    SIGNAL_IDLE = 8;
    FENCE_REQUEST = 16;
    FENCE_RESPONSE = 32;
}

message ClusterMessageHeader {
    ClusterMessageType type = 1;
    int32 message_id = 2;
    string sender_instance_id = 3;
    int32 process_id = 4;
    bool require_ack = 5;
}

message ClusterAck {
    ClusterMessageHeader header = 1;
    int32 ack_message_id = 2;
}

message ClusterSignalHotReload {
    ClusterMessageHeader header = 1;
    string timestamp = 2;
}

message ClusterFenceRequest {
    ClusterMessageHeader header = 1;
    int32 node_id = 2;
}

message ClusterFenceResponse {
    ClusterMessageHeader header = 1;
    int32 node_id = 2;
}

message ClusterAnnounceInstance {
    ClusterMessageHeader header = 1;
    ClusterRole role = 2; 
    bool all_accounted_for = 3;
}

message ClusterSignalIdle {
    ClusterMessageHeader header = 1;
}